lxd/db/openfga.go

```diff
@@ -0,0 +1,610 @@
+		// (Likely an error in the authorization driver).
+		if !requiresProject {
+			return nil, fmt.Errorf("Received unexpected query, entities of type %q do not have a project relation", entityType)
+		}
+
+		// Return a tuple relating the object to its parent project.
+		tuples = []*openfgav1.Tuple{
+			{
+				Key: &openfgav1.TupleKey{
+					Object:   obj,
+					Relation: relation,
+					User:     fmt.Sprintf("%s:%s", entity.TypeProject, entity.ProjectURL(projectName).String()),
+				},
+			},
+		}
+
+	case "server":
+		// If the entity type is project specific but we're looking for server relations then the input is invalid.
+		// (Likely an error in the authorization driver).
+		if requiresProject {
+			return nil, fmt.Errorf("Received unexpected query, entities of type %q do not have a server relation", entityType)
+		}
+
+		// Return a tuple relating the object to the server.
+		tuples = []*openfgav1.Tuple{
+			{
+				Key: &openfgav1.TupleKey{
+					Object:   obj,
+					Relation: relation,
+					User:     fmt.Sprintf("%s:%s", entity.TypeServer, entity.ServerURL().String()),
+				},
+			},
+		}
+
+	default:
+		// Return an error if we get an unexpected relation.
+		return nil, fmt.Errorf("Relation %q not supported", relation)
+	}
+
+	return storage.NewStaticTupleIterator(tuples), nil
+}
+
+// ReadUserTuple reads a single tuple from the store.
+//
+// Observations:
+//   - This method is only called when the `User` field in the given openfgav1.TupleKey is `identity:<identity URL>`.
+//   - Our permissions management system currently does not allow identities to be granted entitlements directly.
+//     Therefore, the only relation that an identity can have is to a `group` via the `member` relation or to a `project` via the
+//     `operator` relation (if the identity is a restricted TLS client). In both of these cases, the tuples have been passed into
+//     the OpenFGA `Check` or `ListObjects` request as contextual tuples.
+//
+// Implementation:
+//   - The tuples that this method is meant to return have been passed in contextually. So validate the input matches
+//     what is expected and return nil.
+func (o *openfgaStore) ReadUserTuple(ctx context.Context, store string, tk *openfgav1.TupleKey) (*openfgav1.Tuple, error) {
+	// Expect the User field to be present.
+	user := tk.GetUser()
+	if user == "" {
+		return nil, fmt.Errorf("ReadUserTuple: User field of tuple key must be provided")
+	}
+
+	// Only allow `identity` for the User type.
+	userEntityType, _, ok := strings.Cut(user, ":")
+	if !ok {
+		return nil, fmt.Errorf("ReadUserTuple: Unexpected format of user field %q", user)
+	}
+
+	if entity.Type(userEntityType) != entity.TypeIdentity {
+		return nil, fmt.Errorf("ReadUserTuple: Entity type %q not supported", userEntityType)
+	}
+
+	return nil, nil
+}
+
+// ReadUsersetTuples is called on check requests. It is used to read all the "users" that have a given relation to
```

Comment:
Could we add a request context optimisation here to only check for the current user rather than checking for all users who have access? (in the future).

Reply:
This is a little nuanced. "user" in this context is not the identity that made the request. OpenFGA tuples are composed of a `user`, a `relation` and an `object`. In the datastore implementation whenever we see a `user` that is an identity, we typically ignore it because all of their relations have been passed in contextually. Here are some examples:

```yaml
user: server:/1.0
relation: server
object: project:/1.0/projects/default
```
```yaml
user: project:/1.0/projects/default
relation: project
object: instance:/1.0/instances/foo?project=default
```
```yaml
user: identity:/1.0/auth/identities/oidc/jane.doe@example.com
relation: member
object: group:/1.0/auth/groups/foo-group
```

We would ignore this one, identities can't be given entitlements directly.
```yaml
user: identity:/1.0/auth/identities/oidc/jane.doe@example.com
relation: can_view
object: instance:/1.0/instances/foo?project=bar
```